### PR TITLE
Addition of file upload component (part 1)

### PR DESCRIPTION
**Pull request by keithtxw**
**Changes**
The addition of the file upload component. Part 1 involves
- Basic structure (drag and drop file ability)
- Rendering of file items
- Error display
- Delete
- Rendering of warning alert

**Action**
- [delete] branch

**Additional information**

- You may refer to this [ticket](https://jira.ship.gov.sg/browse/CCUBE-242)
- [Figma](https://www.figma.com/file/us90jOpWBahsqK2VAluPvD/Flagship-Design-System?type=design&node-id=8137-51564&t=OEgeClHySghxv7oj-0)

### CODE REVIEW

File: src/file-upload/file-item.tsx [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                    </MobileErrorMessage>
+                )}
+            </Content>
+            {renderActionButton()}
+        </Item>
+    );
+};
+
+// =============================================================================
+// HELPER FUNCTIONS
```
Comment (qroll) [inline src/file-upload/file-item.tsx:142]: can this be extracted into a file helper namespace?

File: src/file-upload/file-upload.styles.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { Button } from "../button";
+import { Color } from "../color";
+
+// =============================================================================
+// STYLING
+// =============================================================================
+export const TitleContainer = styled.div`
+    display: flex;
+    flex-direction: column;
+    margin-bottom: 2rem;
```
Comment (qroll) [inline src/file-upload/file-upload.styles.tsx:14]: does margin bottom need to remain when both title and description are not specified?

File: src/file-upload/file-item.tsx [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        if (isLoading) {
+            return (
+                <LoadingActionContainer>
+                    <ProgressBar progress={progress} />
+                </LoadingActionContainer>
+            );
+        } else if (errorMessage) {
+            return (
+                <ItemActionContainer>
+                    <ErrorIconButton
```
Comment (qroll) [inline src/file-upload/file-item.tsx:62]: will need `aria-label`s on these buttons and `aria-hidden` on the children. the aria-label may need to specify the file name too, as the file name doesn't get read.

**Review by qroll — CHANGES_REQUESTED** [review]
referring to FEE's current image upload component, it requires aria-invalid to be set when there are errors

File: src/file-upload/types.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    /** The file object */
+    file: File;
+    /** The alt text or description of the file */
+    description?: string | undefined;
+    /** The upload progress display. Values from 0 to 1 */
+    progress?: number | undefined;
+    /** The error message display to indicate file upload error */
+    errorMessage?: string | undefined;
+    /** If set, the item will be in editable mode */
+    editableMode?: boolean | undefined;
```
[the remark this answers is not in the document]
Comment (keithtxw) [reply src/file-upload/types.ts:12]: Yes the textbox. It's actually used for the other parts. Not relevant at the moment

File: src/file-upload/dropzone.styles.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
+import styled, { css } from "styled-components";
+import { Color } from "../color";
+import { MediaQuery } from "../media";
+import { Text } from "../text";
+import { CloudArrowUpFillIcon } from "@lifesg/react-icons";
```
[the remark this answers is not in the document]
Comment (keithtxw) [reply src/file-upload/dropzone.styles.tsx:5]: funny thing is using the subdirectory import has no savings in the bundle size. I'm intending to abolish the sub directory import for icons

File: src/file-upload/file-item.tsx [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            $loading={isLoading}
+            $editable={isSupportedImageType(file.type)}
+        >
+            <Content>
+                <ItemNameSection>
+                    <ItemText
+                        data-testid="name"
+                        weight={description ? "semibold" : "regular"}
+                    >
+                        {file.name}
```
Comment (qroll) [inline src/file-upload/file-item.tsx:114]: will need to truncate the text if the filename exceeds two lines

File: src/file-upload/dropzone.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    // EVENT HANDLERS
+    // =========================================================================
+    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+        if (event.target.files) {
+            onChange(Array.from(event.target.files));
+        }
+    };
+
+    const handleClick = (event: React.MouseEvent<HTMLInputElement>) => {
+        event.currentTarget.value = "";
```
[the remark this answers is not in the document]
Comment (keithtxw) [reply src/file-upload/dropzone.tsx:61]: Inherited code from FEE. Will find out

File: src/file-upload/file-item.tsx (hunk 6 above)
Comment (keithtxw) [reply src/file-upload/file-item.tsx:114]: Was intending for truncations to be done in the later portions. But i guess i could add them in too

File: src/file-upload/dropzone.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        border,
+        onChange,
+    }: Props,
+    ref: React.Ref<DropzoneElement>
+) => {
+    // =========================================================================
+    // CONST, STATE, REFS
+    // =========================================================================
+    const inputRef = useRef<HTMLInputElement>();
+    const { getRootProps, isDragActive } = useDropzone({
```
[the remark this answers is not in the document]
Comment (qroll) [reply src/file-upload/dropzone.tsx:41]: > on mobile, it should also be set to disabled always

per discussion, there's no need to explicitly disable drag and drop on mobile as the functionality is unlikely to be available. and if the user can get it to work, then we might as well let them proceed 💃

File: src/file-upload/file-upload.styles.tsx (hunk 2 above)
Comment (keithtxw) [reply src/file-upload/file-upload.styles.tsx:14]: Nope. Will conditionally render them

File: src/file-upload/file-item.tsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+                    <ProgressBar progress={progress} />
+                </LoadingActionContainer>
+            );
+        } else if (errorMessage) {
+            return (
+                <ItemActionContainer>
+                    <ErrorIconButton
+                        onClick={handleDelete}
+                        data-testid={`${id}-error-delete-button`}
+                        aria-label={`delete-${name}`}
```
Comment (qroll) [inline src/file-upload/file-item.tsx:105]: voiceover reads it as "delete dash image.png", best to keep as a separate word
```suggestion
                        aria-label={`delete ${name}`}
```

File: src/file-upload/file-upload.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        >
+            <TitleContainer>
+                {title && <Title weight="regular">{title}</Title>}
+                {description && <Description>{description}</Description>}
+            </TitleContainer>
+            {warning && <WarningAlert type="warning">{warning}</WarningAlert>}
+            {renderItems()}
+            <UploadButtonContainer>
+                <UploadButton
+                    id="upload-button"
```
[the remark this answers is not in the document]
Comment (keithtxw) [reply src/file-upload/file-upload.tsx:111]: Oops this is a mistake. Added it previously for another intention. It's no longer needed.

**Review by qroll — APPROVED** [review]

**Pull request merged**